File: scripts/extract_training_data.go
package main

import (
	"bufio"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"strings"
	"time"
)

type Message struct {
	ParentUUID  string                 `json:"parentUuid"`
	SessionID   string                 `json:"sessionId"`
	Type        string                 `json:"type"`
	Message     json.RawMessage        `json:"message"`
	UUID        string                 `json:"uuid"`
	Timestamp   string                 `json:"timestamp"`
	ToolResult  map[string]interface{} `json:"toolUseResult,omitempty"`
}

type MessageContent struct {
	Role    string                   `json:"role"`
	Content interface{}              `json:"content"`
}

type ContentBlock struct {
	Type  string                 `json:"type"`
	Text  string                 `json:"text,omitempty"`
	Name  string                 `json:"name,omitempty"`
	Input map[string]interface{} `json:"input,omitempty"`
}

type ConversationMessage struct {
	Role      string `json:"role"`
	Content   string `json:"content"`
	Timestamp string `json:"timestamp"`
	UUID      string `json:"uuid"`
}

type TrainingItem struct {
	Messages []map[string]string `json:"messages"`
}

type Extractor struct {
	sessions   map[string][]Message
	minTurns   int
	minTokens  int
}

func NewExtractor(minTurns, minTokens int) *Extractor {
	return &Extractor{
		sessions:  make(map[string][]Message),
		minTurns:  minTurns,
		minTokens: minTokens,
	}
}

func (e *Extractor) LoadRawData(filepath string) error {
	start := time.Now()
	fmt.Printf("Loading %s...\n", filepath)

	file, err := os.Open(filepath)
	if err != nil {
		return err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	buf := make([]byte, 0, 64*1024)
	scanner.Buffer(buf, 10*1024*1024) // 10MB buffer for large lines

	lineCount := 0
	for scanner.Scan() {
		line := scanner.Text()
		if strings.TrimSpace(line) == "" {
			continue
		}

		var msg Message
		if err := json.Unmarshal([]byte(line), &msg); err != nil {
			continue // Skip malformed lines
		}

		sessionID := msg.SessionID
		if sessionID == "" {
			sessionID = "unknown"
		}

		e.sessions[sessionID] = append(e.sessions[sessionID], msg)
		lineCount++
	}

	if err := scanner.Err(); err != nil {
		return err
	}

	fmt.Printf("Loaded %d messages from %d sessions in %v\n",
		lineCount, len(e.sessions), time.Since(start))
	return nil
}

func (e *Extractor) extractContent(msg Message) string {
	if msg.Type == "user" {
		var content MessageContent
		if err := json.Unmarshal(msg.Message, &content); err != nil {
			return ""
		}

		// Handle string content
		if str, ok := content.Content.(string); ok {
			return str
		}

		// Handle content blocks
		if blocks, ok := content.Content.([]interface{}); ok {
			var texts []string
			for _, block := range blocks {
				if blockMap, ok := block.(map[string]interface{}); ok {
					if blockMap["type"] == "text" {
						if text, ok := blockMap["text"].(string); ok {
							texts.append(text)
						}
					}
				}
			}
			return strings.Join(texts, "\n")
		}
	} else if msg.Type == "assistant" {
		var content MessageContent
		if err := json.Unmarshal(msg.Message, &content); err != nil {
			return ""
		}

		blocks, ok := content.Content.([]interface{})
		if !ok {
			return ""
		}

		var texts []string
		var toolUses []string

		for _, block := range blocks {
			blockMap, ok := block.(map[string]interface{})
			if !ok {
				continue
			}

			blockType, _ := blockMap["type"].(string)
			switch blockType {
			case "text":
				if text, ok := blockMap["text"].(string); ok {
					texts = append(texts, text)
				}
			case "tool_use":
				if name, ok := blockMap["name"].(string); ok {
					toolUses = append(toolUses, fmt.Sprintf("[TOOL: %s]", name))
				}
			}
		}

		response := strings.Join(texts, "\n")
		if len(toolUses) > 0 {
			response += "\n" + strings.Join(toolUses, "\n")
		}

		return strings.TrimSpace(response)
	}

	return ""
}

func (e *Extractor) ReconstructConversations() [][]ConversationMessage {
	start := time.Now()
	var allConversations [][]ConversationMessage

	for sessionID, messages := range e.sessions {
		_ = sessionID // unused

		// Build message map
		msgMap := make(map[string]Message)
		for _, msg := range messages {
			msgMap[msg.UUID] = msg
		}

		// Find roots
		var roots []Message
		for _, msg := range messages {
			if msg.ParentUUID == "" || msgMap[msg.ParentUUID].UUID == "" {
				roots = append(roots, msg)
			}
		}

		// Build conversations from each root
		for _, root := range roots {
			var conversation []ConversationMessage
			current := root

			for {
				content := e.extractContent(current)
				if content != "" {
					conversation = append(conversation, ConversationMessage{
						Role:      current.Type,
						Content:   content,
						Timestamp: current.Timestamp,
						UUID:      current.UUID,
					})
				}

				// Find next message
				var nextMsg *Message
				for i := range messages {
					if messages[i].ParentUUID == current.UUID {
						nextMsg = &messages[i]
						break
					}
				}

				if nextMsg == nil {
					break
				}
				current = *nextMsg
			}

			if len(conversation) >= e.minTurns {
				allConversations = append(allConversations, conversation)
			}
		}
	}

	fmt.Printf("Reconstructed %d conversations in %v\n",
		len(allConversations), time.Since(start))
	return allConversations
}

func (e *Extractor) FilterQuality(conversations [][]ConversationMessage) [][]ConversationMessage {
	start := time.Now()
	var filtered [][]ConversationMessage

	agenticKeywords := []string{
		"let me", "first", "then", "next", "because", "however",
		"consider", "alternatively", "analysis", "approach",
	}

	for _, conv := range conversations {
		// Check minimum token count
		totalTokens := 0
		for _, msg := range conv {
			totalTokens += len(strings.Fields(msg.Content))
		}
		if totalTokens < e.minTokens {
			continue
		}

		// Check for agentic patterns
		hasTools := false
		hasReasoning := false

		for _, msg := range conv {
			if msg.Role == "assistant" {
				if strings.Contains(msg.Content, "[TOOL:") {
					hasTools = true
				}

				if len(msg.Content) > 200 {
					contentLower := strings.ToLower(msg.Content)
					for _, keyword := range agenticKeywords {
						if strings.Contains(contentLower, keyword) {
							hasReasoning = true
							break
						}
					}
				}
			}
		}

		if hasTools || hasReasoning {
			filtered = append(filtered, conv)
		}
	}

	fmt.Printf("Filtered to %d high-quality conversations in %v\n",
		len(filtered), time.Since(start))
	return filtered
}

func (e *Extractor) FormatForTraining(conversations [][]ConversationMessage) []TrainingItem {
	var trainingData []TrainingItem

	for _, conv := range conversations {
		var messages []map[string]string
		for _, msg := range conv {
			messages = append(messages, map[string]string{
				"role":    msg.Role,
				"content": msg.Content,
			})
		}

		trainingData = append(trainingData, TrainingItem{
			Messages: messages,
		})
	}

	return trainingData
}

func (e *Extractor) SaveTrainingData(outputPath string, trainingData []TrainingItem) error {
	start := time.Now()

	file, err := os.Create(outputPath)
	if err != nil {
		return err
	}
	defer file.Close()

	writer := bufio.NewWriter(file)
	defer writer.Flush()

	totalMessages := 0
	for _, item := range trainingData {
		data, err := json.Marshal(item)
		if err != nil {
			continue
		}
		writer.Write(data)
		writer.WriteByte('\n')
		totalMessages += len(item.Messages)
	}

	// Get file size
	fileInfo, _ := os.Stat(outputPath)
	sizeMB := float64(fileInfo.Size()) / (1024 * 1024)

	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Printf("Training data saved to: %s\n", outputPath)
	fmt.Println(strings.Repeat("=", 60))
	fmt.Printf("Total conversations: %d\n", len(trainingData))
	fmt.Printf("Total messages: %d\n", totalMessages)
	fmt.Printf("File size: %.2f MB\n", sizeMB)
	fmt.Printf("Compression ratio: %.1fx smaller\n", 1700.0/sizeMB)
	fmt.Printf("Processing time: %v\n", time.Since(start))
	fmt.Println(strings.Repeat("=", 60) + "\n")

	return nil
}

func main() {
	inputFile := flag.String("input", "", "Input JSONL file (Claude Code export)")
	outputFile := flag.String("output", "", "Output JSONL file (training format)")
	minTurns := flag.Int("min-turns", 3, "Minimum conversation turns")
	minTokens := flag.Int("min-tokens", 50, "Minimum total tokens")

	flag.Parse()

	if *inputFile == "" || *outputFile == "" {
		fmt.Println("Usage: extract_training_data -input <file> -output <file>")
		flag.PrintDefaults()
		os.Exit(1)
	}

	totalStart := time.Now()

	extractor := NewExtractor(*minTurns, *minTokens)

	if err := extractor.LoadRawData(*inputFile); err != nil {
		fmt.Printf("Error loading data: %v\n", err)
		os.Exit(1)
	}

	conversations := extractor.ReconstructConversations()
	filtered := extractor.FilterQuality(conversations)
	trainingData := extractor.FormatForTraining(filtered)

	if err := extractor.SaveTrainingData(*outputFile, trainingData); err != nil {
		fmt.Printf("Error saving data: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Total execution time: %v\n", time.Since(totalStart))
}

File: scripts/extract_training_data.rs
use serde::{Deserialize, Serialize};
use serde_json::{Value, from_str};
use std::collections::HashMap;
use std::fs::File;
use std::io::{BufRead, BufReader, BufWriter, Write};
use std::time::Instant;
use rayon::prelude::*;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct RawMessage {
    parent_uuid: Option<String>,
    session_id: Option<String>,
    #[serde(rename = "type")]
    msg_type: String,
    message: Option<Value>,
    uuid: String,
    timestamp: Option<String>,
}

#[derive(Debug, Clone)]
struct ConversationMessage {
    role: String,
    content: String,
    timestamp: String,
    uuid: String,
}

#[derive(Debug, Serialize)]
struct TrainingItem {
    m: Vec<HashMap<String, String>>,
}

struct Extractor {
    sessions: HashMap<String, Vec<RawMessage>>,
    min_turns: usize,
    min_tokens: usize,
}

impl Extractor {
    fn new(min_turns: usize, min_tokens: usize) -> Self {
        Self {
            sessions: HashMap::new(),
            min_turns,
            min_tokens,
        }
    }

    fn load_raw_data(&mut self, filepath: &str) -> std::io::Result<()> {
        let start = Instant::now();
        println!("Loading {}...", filepath);

        let file = File::open(filepath)?;
        let reader = BufReader::with_capacity(1024 * 1024, file); // 1MB buffer

        let mut line_count = 0;
        for line in reader.lines() {
            let line = line?;
            if line.trim().is_empty() {
                continue;
            }

            if let Ok(msg) = from_str::<RawMessage>(&line) {
                let session_id = msg.session_id.clone().unwrap_or_else(|| "unknown".to_string());
                self.sessions.entry(session_id).or_insert_with(Vec::new).push(msg);
                line_count += 1;
            }
        }

        println!(
            "Loaded {} messages from {} sessions in {:?}",
            line_count,
            self.sessions.len(),
            start.elapsed()
        );
        Ok(())
    }

    fn extract_content(&self, msg: &RawMessage) -> Option<String> {
        let message = msg.message.as_ref()?;

        match msg.msg_type.as_str() {
            "user" => {
                // Handle user messages
                if let Some(content) = message.get("content") {
                    // String content
                    if let Some(text) = content.as_str() {
                        return Some(text.to_string());
                    }

                    // Array of content blocks
                    if let Some(blocks) = content.as_array() {
                        let texts: Vec<String> = blocks
                            .iter()
                            .filter_map(|block| {
                                if block.get("type")?.as_str()? == "text" {
                                    Some(block.get("text")?.as_str()?.to_string())
                                } else {
                                    None
                                }
                            })
                            .collect();

                        if !texts.is_empty() {
                            return Some(texts.join("\n"));
                        }
                    }
                }
            }
            "assistant" => {
                // Handle assistant messages
                if let Some(content) = message.get("content") {
                    if let Some(blocks) = content.as_array() {
                        let mut texts = Vec::new();
                        let mut tool_uses = Vec::new();

                        for block in blocks {
                            match block.get("type")?.as_str()? {
                                "text" => {
                                    if let Some(text) = block.get("text").and_then(|t| t.as_str()) {
                                        texts.push(text.to_string());
                                    }
                                }
                                "tool_use" => {
                                    if let Some(name) = block.get("name").and_then(|n| n.as_str()) {
                                        tool_uses.push(format!("[TOOL: {}]", name));
                                    }
                                }
                                _ => {}
                            }
                        }

                        let mut response = texts.join("\n");
                        if !tool_uses.is_empty() {
                            response.push('\n');
                            response.push_str(&tool_uses.join("\n"));
                        }

                        if !response.trim().is_empty() {
                            return Some(response);
                        }
                    }
                }
            }
            _ => {}
        }

        None
    }

    fn reconstruct_conversations(&self) -> Vec<Vec<ConversationMessage>> {
        let start = Instant::now();

        let all_conversations: Vec<Vec<ConversationMessage>> = self.sessions
            .par_iter()
            .flat_map(|(_, messages)| {
                // Build message map
                let msg_map: HashMap<&str, &RawMessage> = messages
                    .iter()
                    .map(|msg| (msg.uuid.as_str(), msg))
                    .collect();

                // Find roots
                let roots: Vec<&RawMessage> = messages
                    .iter()
                    .filter(|msg| {
                        msg.parent_uuid.is_none() ||
                        !msg_map.contains_key(msg.parent_uuid.as_ref().unwrap().as_str())
                    })
                    .collect();

                // Build conversations from each root
                roots
                    .into_iter()
                    .filter_map(|root| {
                        let mut conversation = Vec::new();
                        let mut current = root;

                        loop {
                            if let Some(content) = self.extract_content(current) {
                                conversation.push(ConversationMessage {
                                    role: current.msg_type.clone(),
                                    content,
                                    timestamp: current.timestamp.clone().unwrap_or_default(),
                                    uuid: current.uuid.clone(),
                                });
                            }

                            // Find next message
                            let next_msg = messages
                                .iter()
                                .find(|msg| {
                                    msg.parent_uuid.as_ref() == Some(&current.uuid)
                                });

                            match next_msg {
                                Some(msg) => current = msg,
                                None => break,
                            }
                        }

                        if conversation.len() >= self.min_turns {
                            Some(conversation)
                        } else {
                            None
                        }
                    })
                    .collect::<Vec<_>>()
            })
            .collect();

        println!(
            "Reconstructed {} conversations in {:?}",
            all_conversations.len(),
            start.elapsed()
        );

        all_conversations
    }

    fn filter_quality(&self, conversations: Vec<Vec<ConversationMessage>>) -> Vec<Vec<ConversationMessage>> {
        let start = Instant::now();

        let agentic_keywords = [
            "let me", "first", "then", "next", "because", "however",
            "consider", "alternatively", "analysis", "approach",
        ];

        let filtered: Vec<Vec<ConversationMessage>> = conversations
            .into_par_iter()
            .filter(|conv| {
                // Check minimum token count
                let total_tokens: usize = conv
                    .iter()
                    .map(|msg| msg.content.split_whitespace().count())
                    .sum();

                if total_tokens < self.min_tokens {
                    return false;
                }

                // Check for agentic patterns
                let has_tools = conv
                    .iter()
                    .any(|msg| msg.role == "assistant" && msg.content.contains("[TOOL:"));

                let has_reasoning = conv
                    .iter()
                    .any(|msg| {
                        if msg.role == "assistant" && msg.content.len() > 200 {
                            let content_lower = msg.content.to_lowercase();
                            agentic_keywords.iter().any(|kw| content_lower.contains(kw))
                        } else {
                            false
                        }
                    });

                has_tools || has_reasoning
            })
            .collect();

        println!(
            "Filtered to {} high-quality conversations in {:?}",
            filtered.len(),
            start.elapsed()
        );

        filtered
    }

    fn format_for_training(&self, conversations: Vec<Vec<ConversationMessage>>) -> Vec<TrainingItem> {
        conversations
            .into_iter()
            .map(|conv| {
                let messages = conv
                    .into_iter()
                    .map(|msg| {
                        let mut map = HashMap::new();
                        let role = if msg.role == "user" { "u" } else { "a" };
                        map.insert("r".to_string(), role.to_string());
                        map.insert("c".to_string(), msg.content);
                        map
                    })
                    .collect();

                TrainingItem { m: messages }
            })
            .collect()
    }

    fn save_training_data(&self, output_path: &str, training_data: &[TrainingItem]) -> std::io::Result<()> {
        let start = Instant::now();

        let file = File::create(output_path)?;
        let mut writer = BufWriter::with_capacity(1024 * 1024, file); // 1MB buffer

        let mut total_messages = 0;
        for item in training_data {
            serde_json::to_writer(&mut writer, item)?;
            writer.write_all(b"\n")?;
            total_messages += item.m.len();
        }

        writer.flush()?;

        // Get file size
        let metadata = std::fs::metadata(output_path)?;
        let size_mb = metadata.len() as f64 / (1024.0 * 1024.0);

        println!("\n{}", "=".repeat(60));
        println!("Training data saved to: {}", output_path);
        println!("{}", "=".repeat(60));
        println!("Total conversations: {}", training_data.len());
        println!("Total messages: {}", total_messages);
        println!("File size: {:.2} MB", size_mb);
        println!("Compression ratio: {:.1}x smaller", 1700.0 / size_mb);
        println!("Processing time: {:?}", start.elapsed());
        println!("{}\n", "=".repeat(60));

        Ok(())
    }
}

fn main() -> std::io::Result<()> {
    let args: Vec<String> = std::env::args().collect();

    if args.len() < 3 {
        eprintln!("Usage: {} <input.jsonl> <output.jsonl> [min_turns] [min_tokens]", args[0]);
        std::process::exit(1);
    }

    let input_file = &args[1];
    let output_file = &args[2];
    let min_turns = args.get(3).and_then(|s| s.parse().ok()).unwrap_or(3);
    let min_tokens = args.get(4).and_then(|s| s.parse().ok()).unwrap_or(50);

    let total_start = Instant::now();

    let mut extractor = Extractor::new(min_turns, min_tokens);

    extractor.load_raw_data(input_file)?;
    let conversations = extractor.reconstruct_conversations();
    let filtered = extractor.filter_quality(conversations);
    let training_data = extractor.format_for_training(filtered);
    extractor.save_training_data(output_file, &training_data)?;

    println!("Total execution time: {:?}", total_start.elapsed());

    Ok(())
}
